File: 0205-isomorphic-strings/0205-isomorphic-strings.py
## time: o(n)
## space : o(n)
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        mapST, mapTS = {}, {}  # Create two hash maps for character mappings
        
        for i in range(len(s)):  # Iterate through the strings
            c1, c2 = s[i], t[i]  # Get the current characters from s and t
            
            # Check if the current mappings violate isomorphism
            if (c1 in mapST and mapST[c1] != c2) or (c2 in mapTS and mapTS[c2] != c1):
                return False
            
            # Update the hash maps with the mappings from s to t and t to s
            mapST[c1] = c2
            mapTS[c2] = c1
            
        return True  # If all characters are validly mapped, return True

File: 0205-isomorphic-strings/0205-isomorphic-strings.cpp
/*Given two strings s and t, determine if they are isomorphic.

Two strings s and t are isomorphic if the characters in s can be replaced to get t.

All occurrences of a character must be replaced with another character while preserving the order of characters. No two characters may map to the same character, but a character may map to itself.

 */
class Solution {
public:
    bool isIsomorphic(string s, string t) {
       unordered_map<char,vector<int>>m1;
         unordered_map<char,vector<int>>m2;
        for(int i=0;i<s.length();i++){
            m1[s[i]].push_back(i);
             m2[t[i]].push_back(i);
            
            if(m1[s[i]]!=m2[t[i]])
                return false;
        }
        return true;
        
    }
};

/*int main() {
    Solution solution;

    // Test case 1
    string s1 = "egg";
    string t1 = "add";
    cout << solution.isIsomorphic(s1, t1) << endl; // Output: true

    // Test case 2
    string s2 = "foo";
    string t2 = "bar";
    cout << solution.isIsomorphic(s2, t2) << endl; // Output: false

    // Test case 3
    string s3 = "paper";
    string t3 = "title";
    cout << solution.isIsomorphic(s3, t3) << endl; // Output: true

    return 0;
}
*/
